File: src/Functional/Elixir/factory_method.exs
defmodule FactoryMethod do
  def create_postgres do
    %{
      connect: fn -> IO.puts("PostgreSQL connect") end,
      query: fn -> IO.puts("PostgreSQL query") end
    }
  end

  def create_mysql do
    %{
      connect: fn -> IO.puts("MySQL connect") end,
      query: fn -> IO.puts("MySQL query") end
    }
  end

  def use_database(create_database) do
    database = create_database.()
    database.connect.()
    database.query.()
  end
end

FactoryMethod.use_database(&FactoryMethod.create_postgres/0)
FactoryMethod.use_database(&FactoryMethod.create_mysql/0)

File: src/Functional/Erlang/factory_method.erl
-module(factory_method).
-export([main/0]).

create_postgres() ->
    #{connect => fun() -> io:format("PostgreSQL connect~n") end,
      query => fun() -> io:format("PostgreSQL query~n") end}.

create_mysql() ->
    #{connect => fun() -> io:format("MySQL connect~n") end,
      query => fun() -> io:format("MySQL query~n") end}.

use_database(CreateDatabase) ->
    Database = CreateDatabase(),
    Connect = maps:get(connect, Database),
    Query = maps:get(query, Database),
    Connect(),
    Query().

main() ->
    use_database(fun create_postgres/0),
    use_database(fun create_mysql/0).
